tests/e2e/refcount_test.go

```diff
@@ -0,0 +1,150 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This test suite includes refcount test on plugin side
+
+// +build runonce
+
+package e2e
+
+import (
+	"strconv"
+	"strings"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/vmware/docker-volume-vsphere/tests/utils/dockercli"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/inputparams"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/misc"
+	"github.com/vmware/docker-volume-vsphere/tests/utils/verification"
+)
+
+type RefcountTestSuite struct {
+	config *inputparams.TestConfig
+
+	volumeName        string
+	containerNameList []string
+}
+
+func (s *RefcountTestSuite) SetUpSuite(c *C) {
+	s.config = inputparams.GetTestConfig()
+	if s.config == nil {
+		c.Skip("Unable to retrieve test config, skipping refcount tests")
+	}
+}
+func (s *RefcountTestSuite) SetUpTest(c *C) {
+	s.volumeName = inputparams.GetUniqueVolumeName("RefcountTest")
+
+	// Create a volume
+	out, err := dockercli.CreateVolume(s.config.DockerHosts[0], s.volumeName)
+	c.Assert(err, IsNil, Commentf(out))
+}
+
+func (s *RefcountTestSuite) TearDownTest(c *C) {
+
+	// If containers are leftover due to test fail, clean here
+	cName := strings.Join(s.containerNameList, " ")
+	if cName != "" {
+		out, err := dockercli.RemoveContainer(s.config.DockerHosts[0], cName)
+		c.Assert(err, IsNil, Commentf(out))
+	}
+}
+
+var _ = Suite(&RefcountTestSuite{})
+
+// Test for refcount recovery
+// * Create a volume
+// * Attach 5 containers to it with restart flag
+// * Verify attached status after each attach
+// * Check the refcount log in log file
+// * Verify the content on volume created by containers
+// * Check the volume entry on mount file : /proc/mounts
+// * Try to delete volume. Should Fail
+// * Kill Docker (ungraceful)
+// * Check the recovery log record in the log file
+// * Cleanup the containers
+// * Delete the volume. Should succeed
+// * There should be no entry in /proc/mounts
+
+func (s *RefcountTestSuite) TestRefcountingRecovery(c *C) {
+	misc.LogTestStart(c.TestName())
+
+	// Check if volume is available in docker host
+	accessible := verification.CheckVolumeAvailability(s.config.DockerHosts[0], s.volumeName)
+	c.Assert(accessible, Equals, true, Commentf("Volume %s is not available", s.volumeName))
+
+	numContainers := 5
+
+	// Each container creates a file with its index as the filename suffix. After that it keeps running.
+	for i := 0; i < numContainers; i++ {
+		cName := inputparams.GetUniqueContainerName(c.TestName())
+		out, err := dockercli.AttachWriteVolumeWithRestart(s.config.DockerHosts[0], s.volumeName, cName, "file_"+strconv.Itoa(i))
+		c.Assert(err, IsNil, Commentf(out))
+
+		s.containerNameList = append(s.containerNameList, cName)
+
+		status := verification.VerifyAttachedStatus(s.volumeName, s.config.DockerHosts[0], s.config.EsxHost)
```

Comment:
why are we verifying status `numContainers ` times? isn't it sufficient to verify once after the loop?